lib/user/store.js

```diff
@@ -0,0 +1,170 @@
+var q = require('q');
+q.longStackSupport = process.env.NODE_ENV === 'development';
+
+var verifyErrMessage = 'User not found with supplied credentials';
+
+function findById(id, data) {
+  return _.find(data, {id: id});
+}
+function findByUsername(username, data) {
+  return _.find(data, {username: username});
+}
+
+function cloneAndCleanup(user) {
+  var cloned = _.clone(user);
+  delete cloned.password;
+  return cloned;
+}
+
+function setUserPassword(user, pwd, cb) {
+  hash.saltAndHash(pwd, function(err, hashed) {
+    if (err) {
+      return cb(err);
+    }
+    user.password = hashed;
+    cb(null, user);
+  });
+}
+
+function Store(datasetId, data) {
+  this.data = data;
+  this.topic = {};
+  this.subscription = {};
+  this.datasetId = datasetId;
+}
+
+/**
+ * Helper function to reset underlying collection for testing
+ */
+Store.prototype.setAll = function(arr) {
+  this.data = arr;
+};
+
+Store.prototype.list = ArrayStore.prototype.list;
+
+Store.prototype.read = function(id) {
+  var user = findById(id, this.data);
+  if (!user) {
+    return q.reject(new Error('User not found'));
+  }
+  return q(cloneAndCleanup(user));
+};
+
+Store.prototype.byUsername = function(username) {
+  var user = findByUsername(username, this.data);
+  if (!user) {
+    return q.reject(new Error('User not found'));
+  }
+  return q(cloneAndCleanup(user));
+};
+
+
+Store.prototype.create = function(user) {
+  user = _.clone(user);
+  user.id = shortid.generate();
+  var self = this;
+  return q.nfcall(setUserPassword, user, user.password).then(function(user) {
+    self.data.push(user);
+    return cloneAndCleanup(user);
+  });
+};
+
+Store.prototype.update = function(user) {
+  var originalUser = findById(user.id, this.data);
+  if (!originalUser) {
+    return q.reject(new Error('User not found'));
+  }
+  delete user.id;
+  _.assign(originalUser, user);
+  // update password since this should only be called from the portal
+  var passwordPromise = user.password ? q.nfcall(setUserPassword, originalUser, user.password) : q(originalUser);
+  return passwordPromise.then(cloneAndCleanup);
+};
+
+Store.prototype.updatePassword = function(username, oldPwd, newPwd) {
+  var user = findByUsername(username, this.data);
+  if (!user) {
+    return q.reject(new Error(verifyErrMessage));
+  }
+  return q.nfcall(hash.verify, oldPwd, user.password).then(function(match) {
+    if (!match) {
+      throw new Error();
+    }
+    return q.nfcall(setUserPassword, user, newPwd);
+  }).then(function(user) {
+    return cloneAndCleanup(user);
+  }).catch(function() {
+    return q.reject(new Error(verifyErrMessage));
+  });
+};
+
+function doVerify(password, user) {
+  // user.password here should be the stored hash
+  return q.nfcall(hash.verify, password, user.password).then(function(match) {
+    if (!match) {
+      throw new Error(verifyErrMessage);
+    } else {
+      return true;
+    }
+  });
+}
+
+Store.prototype.verifyPassword = function(username, password) {
```

Comment:
Doing this atm. Just wanted to point out that I removed an exponential increase in the delay used to verify a password in favor of a fixed delay of .5s, because mediator topics have a timeout of 2s, we would quickly overshoot it and cause problems in the upper layers.

I don't think the delay should be added to the jsdocs though since it's an implementation detail that's not relevant to client code.